sklearn/neighbors/nca.py

```diff
@@ -0,0 +1,596 @@
+# coding: utf-8
+"""
+Neighborhood Component Analysis
+"""
+
+# License: BSD 3 Clause
+
+from __future__ import print_function
+
+from warnings import warn
+import numpy as np
+import sys
+import time
+from scipy.optimize import minimize
+from ..utils.fixes import logsumexp
+from ..metrics import pairwise_distances
+from ..base import BaseEstimator, TransformerMixin
+from ..preprocessing import LabelEncoder
+from ..decomposition import PCA
+from ..utils.multiclass import check_classification_targets
+from ..utils.random import check_random_state
+from ..utils.validation import check_is_fitted, check_array, check_X_y
+from ..externals.six import integer_types
+from ..exceptions import ConvergenceWarning
+
+
+class NeighborhoodComponentsAnalysis(BaseEstimator, TransformerMixin):
+    """Neighborhood Components Analysis
+
+    Parameters
+    ----------
+    n_components : int, optional (default=None)
+        Preferred dimensionality of the embedding.
+        If None it is inferred from ``init``.
+
+    init : string or numpy array, optional (default='auto')
+        Initialization of the linear transformation. Possible options are
+        'auto', 'pca', 'lda', 'identity', 'random', and a numpy array of shape
+        (n_features_a, n_features_b).
+
+        'auto':
+            Depending on ``n_components``, the most reasonable initialization
+            will be chosen among the following ones. First, we try to use
+            'lda', as it uses labels information: if ``n_components <=
+            n_classes``, ``init='lda'``. If we can't, we then try 'pca', as it
+            projects data in meaningful directions (those of higher variance):
+            if ``n_components < min(n_features, n_samples)``, ``init = 'pca'``.
+            Otherwise, we just use 'identity'.
```

Comment:
Good question, I did not think this through... Looking at the code of `_initialize` not always: the only possible case where it would not be identity would be if `n_classes - 1 >= n_features`. In this case the init would be 'lda', which I think is a reasonable init in this case. In fact the reasoning from the code is just that if `n_components` is None, then we set `n_components` to `n_features` and then we choose the appropriate auto init. Therefore I'll replace indeed this:

```python
    n_components : int, optional (default=None)
        Preferred dimensionality of the embedding.
        If None it is inferred from ``init``.
```

by this

```python
    n_components : int, optional (default=None)
        Preferred dimensionality of the embedding.
        If None it will be set to ``n_features``.
```

Tell me if you think of something else

Reply:
This looks good. We should test this in the tests, though.

Reply:
the above docstring can be simplified to be clearer.

```
If n_components < n_features, 'lda' init is used.
If XXX then XXX is used.
...
```

currently it's a bit convoluted